crates/cubecl-wgpu/src/compute/server.rs

```diff
@@ -267,30 +283,30 @@ impl WgpuServer {
                 mapped_at_creation: false,
             });
 
-            self.encoder
-                .resolve_query_set(&self.query_set, 0..2, &resolved, 0);
+            self.encoder.resolve_query_set(queries, 0..2, &resolved, 0);
+            let fut = self.read_wgpu_buffer(&resolved, 0, size);
 
-            let period = self.queue.get_timestamp_period() as f64 * 1e-9;
-            Some((self.read_wgpu_buffer(&resolved, 0, size), period))
-        } else {
-            None
-        };
-
-        self.query_started = false;
-        let duration_profiled = self.duration_profiled;
-
-        async move {
-            if let Some((fut, period)) = fut {
+            Box::pin(async move {
                 let data = fut
                     .await
                     .chunks_exact(8)
                     .map(|x| u64::from_le_bytes(x.try_into().unwrap()))
                     .collect::<Vec<_>>();
                 let delta = data[1] - data[0];
-                Duration::from_secs_f64(delta as f64 * period) + duration_profiled
-            } else {
-                duration_profiled
-            }
+                Duration::from_secs_f64(delta as f64 * period)
+            })
+        } else {
+            // TODO: This should work queue.on_submitted_work_done() but that
+            // is not yet implemented on wgpu https://github.com/gfx-rs/wgpu/issues/6395
+            //
+            // For now, instead do a dummy readback. This *seems* to wait for the entire
+            // queue to be done.
+            let dummy = self.empty(32);
+            let fut = self.read(dummy.binding());
+            Box::pin(async move {
+                fut.await;
+                start_time.elapsed()
+            })
```

Comment:
Could we simply `poll::wait` the device here instead?

Reply:
Maybe but then this path doesn't for WebGPU again, and the timestamp queries extension is technically optional, so you could hit this.

on_submitted_work_done is the real fix, hope wgpu has it soon.